sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala

```diff
@@ -351,8 +351,13 @@ case class AlterTableAddPartitionCommand(
         "ALTER TABLE ADD PARTITION is not allowed for tables defined using the datasource API")
     }
     val parts = partitionSpecsAndLocs.map { case (spec, location) =>
+      val normalizedSpec = PartitioningUtils.normalizePartitionSpec(
+        spec,
+        table.partitionColumnNames,
+        table.identifier.quotedString,
+        sparkSession.sessionState.conf.resolver)
       // inherit table storage format (possibly except for location)
-      CatalogTablePartition(spec, table.storage.copy(locationUri = location))
+      CatalogTablePartition(normalizedSpec, table.storage.copy(locationUri = location))
```

Comment:
Btw, I noticed a number of missing case conversions when combining this with https://github.com/apache/spark/pull/15515/files. They seem to be mostly from CatalogTablePartitions listed from the hive metastore.

Since this is hard to test until that is merged, maybe add a TODO in `HiveClientImpl:836` to do this normalization for partition specs read from the metastore?